### DISCUSSION TITLE
Functions must begin with :or, or :all (=) to receive input

### DISCUSSION THREAD

**Discussion by btheado** [TiddlyWiki/TiddlyWiki5#8786] [General]
Based on experiments, I find that functions must begin with `:or` (or blank), or `:all` (or `=`) prefix in order to receive input. Given that constraint, functions which need to process input should using one of the other filter run prefixes should start always start with `[all[]]` , or `:or[all[]]`, or `:all[all[]]`, or `=[all[]]`. The `all` operator serves as an identity/no-op/passthrough operator. There are other operators which can act as passthrough (i.e. `[prefix[]]` with no argument), but `[all[]]` might be the most clear.

I did not find any mention of this requirement in the docs, so I plan to add it. But one thing I don't know is the reason for it. Does it have to be this way? I could just document the requirement as a fact, but it would be nice to include an explanation. Can anyone help with the explanation?

Here is a passing test case which shows any named prefix can receive input as long as it starts with `[all[]]`

```
title: Functions/FunctionNamedFilterRunPrefix1
description: Functions with named prefixes receive input using [all[]]
type: text/vnd.tiddlywiki-multiple
tags: [[$:/tags/wiki-test-spec]]

title: Output

\whitespace trim
\function all.identity()                  :all[all[]]
\function and.identity()          [all[]] :and[all[]]
\function cascade.identity()      [all[]] :cascade[all[]]
\function else.one()              [all[]] :else[[1]]
\function except.two()            [all[]] :except[[2]]
\function filter.identity()       [all[]] :filter[all[]]
\function intersection.identity() [all[]] :intersection[all[]]
\function map.identity()          [all[]] :map[all[]]
\function or.identity()                   :or[all[]]
\function reduce.identity()       [all[]] :reduce[all[]]
\function sort.identity()         [all[]] :sort[all[]]
\function then.identity()         [all[]] :then[all[]]

<$text text={{{ 
  [title[all.identity]all.identity[]]
  [title[and.identity]and.identity[]]
  [title[cascade.identity]cascade.identity[]]
  [title[else.one]else.one[]]
  [enlist[2 except.two]except.two[]]
  [title[filter.identity]filter.identity[]]
  [title[intersection.identity]intersection.identity[]]
  [title[map.identity]map.identity[]]
  [title[or.identity]or.identity[]]
  [title[reduce.identity]reduce.identity[]]
  [title[sort.identity]sort.identity[]]
  [title[then.identity]then.identity[]]
  :and[join[ ]]
}}}/>

+
title: ExpectedResult

all.identity and.identity cascade.identity else.one except.two filter.identity intersection.identity map.identity or.identity reduce.identity sort.identity then.identity
```

Here is a test case which shows without the `[all[]]`, most of the prefixes don't receive input

```
title: Functions/FunctionNameFilterRunPrefix2
description: Only functions starting with :or or :all named prefixes receive input
type: text/vnd.tiddlywiki-multiple
tags: [[$:/tags/wiki-test-spec]]

title: Output

\whitespace trim
\function all.identity()          :all[all[]]
\function and.identity()          :and[all[]]
\function cascade.identity()      :cascade[all[]]
\function else.one()              :else[[1]]
\function except.two()            :except[[2]]
\function filter.identity()       :filter[all[]]
\function intersection.identity() :intersection[all[]]
\function map.identity()          :map[all[]]
\function or.identity()           :or[all[]]
\function reduce.identity()       :reduce[all[]]
\function sort.identity()         :sort[all[]]
\function then.identity()         :then[all[]]

<$text text={{{ 
  [title[all.identity]all.identity[]]
  [title[and.identity]and.identity[]]
  [title[cascade.identity]cascade.identity[]]
  [title[else.one]else.one[]]
  [enlist[2 except.two]except.two[]]
  [title[filter.identity]filter.identity[]]
  [title[intersection.identity]intersection.identity[]]
  [title[map.identity]map.identity[]]
  [title[or.identity]or.identity[]]
  [title[reduce.identity]reduce.identity[]]
  [title[sort.identity]sort.identity[]]
  [title[then.identity]then.identity[]]
  :and[join[ ]]
}}}/>

+
title: ExpectedResult

all.identity 1 or.identity
```

I wrote similar test cases for the subfilter operator and the behavior is consistent.

Check out a [live version at the share site](https://tiddlywiki.com/prerelease/share##%5B%7B%22title%22%3A%22%24%3A%2Fstate%2Fhttp-requests%22%2C%22text%22%3A%220%22%7D%2C%7B%22created%22%3A%2220241126205809402%22%2C%22title%22%3A%22%24%3A%2Fstate%2Fsidebar%22%2C%22text%22%3A%22no%22%2C%22modified%22%3A%2220241126205809402%22%7D%2C%7B%22title%22%3A%22%24%3A%2FStoryList%22%2C%22created%22%3A%2220241126205511401%22%2C%22text%22%3A%22%22%2C%22list%22%3A%22Functions%2FFunctionNamedFilterRunPrefix1%20Functions%2FFunctionNamedFilterRunPrefix2%20Subfilters%2FSubfilterNamedFilterRunPrefix1%20Subfilters%2FSubfilterNamedFilterRunPrefix2%22%2C%22modified%22%3A%2220241126205749698%22%7D%2C%7B%22title%22%3A%22Functions%2FFunctionNamedFilterRunPrefix1%22%2C%22description%22%3A%22Functions%20with%20named%20prefixes%20receive%20input%20using%20%5Ball%5B%5D%5D%22%2C%22type%22%3A%22text%2Fvnd.tiddlywiki-multiple%22%2C%22tags%22%3A%22%24%3A%2Ftags%2Fwiki-test-spec%22%2C%22text%22%3A%22title%3A%20Output%5Cn%5Cn%5C%5Cwhitespace%20trim%5Cn%5C%5Cfunction%20all.identity%28%29%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%3Aall%5Ball%5B%5D%5D%5Cn%5C%5Cfunction%20and.identity%28%29%20%20%20%20%20%20%20%20%20%20%5Ball%5B%5D%5D%20%3Aand%5Ball%5B%5D%5D%5Cn%5C%5Cfunction%20cascade.identity%28%29%20%20%20%20%20%20%5Ball%5B%5D%5D%20%3Acascade%5Ball%5B%5D%5D%5Cn%5C%5Cfunction%20else.one%28%29%20%20%20%20%20%20%20%20%20%20%20%20%20%20%5Ball%5B%5D%5D%20%3Aelse%5B%5B1%5D%5D%5Cn%5C%5Cfunction%20except.two%28%29%20%20%20%20%20%20%20%20%20%20%20%20%5Ball%5B%5D%5D%20%3Aexcept%5B%5B2%5D%5D%5Cn%5C%5Cfunction%20filter.identity%28%29%20%20%20%20%20%20%20%5Ball%5B%5D%5D%20%3Afilter%5Ball%5B%5D%5D%5Cn%5C%5Cfunction%20intersection.identity%28%29%20%5Ball%5B%5D%5D%20%3Aintersection%5Ball%5B%5D%5D%5Cn%5C%5Cfunction%20map.identity%28%29%20%20%20%20%20%20%20%20%20%20%5Ball%5B%5D%5D%20%3Amap%5Ball%5B%5D%5D%5Cn%5C%5Cfunction%20or.identity%28%29%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%3Aor%5Ball%5B%5D%5D%5Cn%5C%5Cfunction%20reduce.identity%28%29%20%20%20%20%20%20%20%5Ball%5B%5D%5D%20%3Areduce%5Ball%5B%5D%5D%5Cn%5C%5Cfunction%20sort.identity%28%29%20%20%20%20%20%20%20%20%20%5Ball%5B%5D%5D%20%3Asort%5Ball%5B%5D%5D%5Cn%5C%5Cfunction%20then.identity%28%29%20%20%20%20%20%20%20%20%20%5Ball%5B%5D%5D%20%3Athen%5Ball%5B%5D%5D%5Cn%5Cn%3C%24text%20text%3D%7B%7B%7B%20%5Cn%20%20%5B%5Ball.identity%5Dall.identity%5B%5D%5D%5Cn%20%20%5B%5Band.identity%5Dand.identity%5B%5D%5D%5Cn%20%20%5B%5Bcascade.identity%5Dcascade.identity%5B%5D%5D%5Cn%20%20%5B%5Belse.one%5Delse.one%5B%5D%5D%5Cn%20%20%5Benlist%5B2%20except.two%5Dexcept.two%5B%5D%5D%5Cn%20%20%5B%5Bfilter.identity%5Dfilter.identity%5B%5D%5D%5Cn%20%20%5B%5Bintersection.identity%5Dintersection.identity%5B%5D%5D%5Cn%20%20%5B%5Bmap.identity%5Dmap.identity%5B%5D%5D%5Cn%20%20%5B%5Bor.identity%5Dor.identity%5B%5D%5D%5Cn%20%20%5B%5Breduce.identity%5Dreduce.identity%5B%5D%5D%5Cn%20%20%5B%5Bsort.identity%5Dsort.identity%5B%5D%5D%5Cn%20%20%5B%5Bthen.identity%5Dthen.identity%5B%5D%5D%5Cn%20%20%3Aand%5Bjoin%5B%20%5D%5D%5Cn%7D%7D%7D%2F%3E%5Cn%5Cn%2B%5Cntitle%3A%20ExpectedResult%5Cn%5Cnall.identity%20and.identity%20cascade.identity%20else.one%20except.two%20filter.identity%20intersection.identity%20map.identity%20or.identity%20reduce.identity%20sort.identity%20then.identity%22%7D%2C%7B%22created%22%3A%2220241126205419148%22%2C%22text%22%3A%22title%3A%20Output%5Cn%5Cn%5C%5Cwhitespace%20trim%5Cn%5C%5Cfunction%20all.identity%28%29%20%20%20%20%20%20%20%20%20%20%3Aall%5Ball%5B%5D%5D%5Cn%5C%5Cfunction%20and.identity%28%29%20%20%20%20%20%20%20%20%20%20%3Aand%5Ball%5B%5D%5D%5Cn%5C%5Cfunction%20cascade.identity%28%29%20%20%20%20%20%20%3Acascade%5Ball%5B%5D%5D%5Cn%5C%5Cfunction%20else.one%28%29%20%20%20%20%20%20%20%20%20%20%20%20%20%20%3Aelse%5B%5B1%5D%5D%5Cn%5C%5Cfunction%20except.two%28%29%20%20%20%20%20%20%20%20%20%20%20%20%3Aexcept%5B%5B2%5D%5D%5Cn%5C%5Cfunction%20filter.identity%28%29%20%20%20%20%20%20%20%3Afilter%5Ball%5B%5D%5D%5Cn%5C%5Cfunction%20intersection.identity%28%29%20%3Aintersection%5Ball%5B%5D%5D%5Cn%5C%5Cfunction%20map.identity%28%29%20%20%20%20%20%20%20%20%20%20%3Amap%5Ball%5B%5D%5D%5Cn%5C%5Cfunction%20or.identity%28%29%20%20%20%20%20%20%20%20%20%20%20%3Aor%5Ball%5B%5D%5D%5Cn%5C%5Cfunction%20reduce.identity%28%29%20%20%20%20%20%20%20%3Areduce%5Ball%5B%5D%5D%5Cn%5C%5Cfunction%20sort.identity%28%29%20%20%20%20%20%20%20%20%20%3Asort%5Ball%5B%5D%5D%5Cn%5C%5Cfunction%20then.identity%28%29%20%20%20%20%20%20%20%20%20%3Athen%5Ball%5B%5D%5D%5Cn%5Cn%3C%24text%20text%3D%7B%7B%7B%20%5Cn%20%20%5B%5Ball.identity%5Dall.identity%5B%5D%5D%5Cn%20%20%5B%5Band.identity%5Dand.identity%5B%5D%5D%5Cn%20%20%5B%5Bcascade.identity%5Dcascade.identity%5B%5D%5D%5Cn%20%20%5B%5Belse.one%5Delse.one%5B%5D%5D%5Cn%20%20%5Benlist%5B2%20except.two%5Dexcept.two%5B%5D%5D%5Cn%20%20%5B%5Bfilter.identity%5Dfilter.identity%5B%5D%5D%5Cn%20%20%5B%5Bintersection.identity%5Dintersection.identity%5B%5D%5D%5Cn%20%20%5B%5Bmap.identity%5Dmap.identity%5B%5D%5D%5Cn%20%20%5B%5Bor.identity%5Dor.identity%5B%5D%5D%5Cn%20%20%5B%5Breduce.identity%5Dreduce.identity%5B%5D%5D%5Cn%20%20%5B%5Bsort.identity%5Dsort.identity%5B%5D%5D%5Cn%20%20%5B%5Bthen.identity%5Dthen.identity%5B%5D%5D%5Cn%20%20%3Aand%5Bjoin%5B%20%5D%5D%5Cn%7D%7D%7D%2F%3E%5Cn%5Cn%2B%5Cntitle%3A%20ExpectedResult%5Cn%5Cnall.identity%201%20or.identity%22%2C%22title%22%3A%22Functions%2FFunctionNamedFilterRunPrefix2%22%2C%22description%22%3A%22Only%20functions%20starting%20with%20%3Aor%20or%20%3Aall%20named%20prefixes%20receive%20input%22%2C%22type%22%3A%22text%2Fvnd.tiddlywiki-multiple%22%2C%22tags%22%3A%22%24%3A%2Ftags%2Fwiki-test-spec%22%2C%22modified%22%3A%2220241126205420713%22%7D%2C%7B%22title%22%3A%22Subfilters%2FSubfilterNamedFilterRunPrefix1%22%2C%22description%22%3A%22Subfilters%20with%20named%20prefixes%20receive%20input%20using%20%5Ball%5B%5D%5D%22%2C%22type%22%3A%22text%2Fvnd.tiddlywiki-multiple%22%2C%22tags%22%3A%22%24%3A%2Ftags%2Fwiki-test-spec%22%2C%22text%22%3A%22title%3A%20Output%5Cn%5Cn%5C%5Cwhitespace%20trim%5Cn%5C%5Cprocedure%20all.identity%28%29%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%3Aall%5Ball%5B%5D%5D%5Cn%5C%5Cprocedure%20and.identity%28%29%20%20%20%20%20%20%20%20%20%20%5Ball%5B%5D%5D%20%3Aand%5Ball%5B%5D%5D%5Cn%5C%5Cprocedure%20cascade.identity%28%29%20%20%20%20%20%20%5Ball%5B%5D%5D%20%3Acascade%5Ball%5B%5D%5D%5Cn%5C%5Cprocedure%20else.one%28%29%20%20%20%20%20%20%20%20%20%20%20%20%20%20%5Ball%5B%5D%5D%20%3Aelse%5B%5B1%5D%5D%5Cn%5C%5Cprocedure%20except.two%28%29%20%20%20%20%20%20%20%20%20%20%20%20%5Ball%5B%5D%5D%20%3Aexcept%5B%5B2%5D%5D%5Cn%5C%5Cprocedure%20filter.identity%28%29%20%20%20%20%20%20%20%5Ball%5B%5D%5D%20%3Afilter%5Ball%5B%5D%5D%5Cn%5C%5Cprocedure%20intersection.identity%28%29%20%5Ball%5B%5D%5D%20%3Aintersection%5Ball%5B%5D%5D%5Cn%5C%5Cprocedure%20map.identity%28%29%20%20%20%20%20%20%20%20%20%20%5Ball%5B%5D%5D%20%3Amap%5Ball%5B%5D%5D%5Cn%5C%5Cprocedure%20or.identity%28%29%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%3Aor%5Ball%5B%5D%5D%5Cn%5C%5Cprocedure%20reduce.identity%28%29%20%20%20%20%20%20%20%5Ball%5B%5D%5D%20%3Areduce%5Ball%5B%5D%5D%5Cn%5C%5Cprocedure%20sort.identity%28%29%20%20%20%20%20%20%20%20%20%5Ball%5B%5D%5D%20%3Asort%5Ball%5B%5D%5D%5Cn%5C%5Cprocedure%20then.identity%28%29%20%20%20%20%20%20%20%20%20%5Ball%5B%5D%5D%20%3Athen%5Ball%5B%5D%5D%5Cn%5Cn%3C%24text%20text%3D%7B%7B%7B%20%5Cn%20%20%5B%5Ball.identity%5Dsubfilter%3Call.identity%3E%5D%5Cn%20%20%5B%5Band.identity%5Dsubfilter%3Cand.identity%3E%5D%5Cn%20%20%5B%5Bcascade.identity%5Dsubfilter%3Ccascade.identity%3E%5D%5Cn%20%20%5B%5Belse.one%5Dsubfilter%3Celse.one%3E%5D%5Cn%20%20%5Benlist%5B2%20except.two%5Dsubfilter%3Cexcept.two%3E%5D%5Cn%20%20%5B%5Bfilter.identity%5Dsubfilter%3Cfilter.identity%3E%5D%5Cn%20%20%5B%5Bintersection.identity%5Dsubfilter%3Cintersection.identity%3E%5D%5Cn%20%20%5B%5Bmap.identity%5Dsubfilter%3Cmap.identity%3E%5D%5Cn%20%20%5B%5Bor.identity%5Dsubfilter%3Cor.identity%3E%5D%5Cn%20%20%5B%5Breduce.identity%5Dsubfilter%3Creduce.identity%3E%5D%5Cn%20%20%5B%5Bsort.identity%5Dsubfilter%3Csort.identity%3E%5D%5Cn%20%20%5B%5Bthen.identity%5Dsubfilter%3Cthen.identity%3E%5D%5Cn%20%20%3Aand%5Bjoin%5B%20%5D%5D%5Cn%7D%7D%7D%2F%3E%5Cn%5Cn%2B%5Cntitle%3A%20ExpectedResult%5Cn%5Cnall.identity%20and.identity%20cascade.identity%20else.one%20except.two%20filter.identity%20intersection.identity%20map.identity%20or.identity%20reduce.identity%20sort.identity%20then.identity%22%7D%2C%7B%22title%22%3A%22Subfilters%2FSubfilterNamedFilterRunPrefix2%22%2C%22description%22%3A%22Only%20subfilters%20starting%20with%20%3Aor%20or%20%3Aall%20named%20prefixes%20receive%20input%22%2C%22type%22%3A%22text%2Fvnd.tiddlywiki-multiple%22%2C%22tags%22%3A%22%24%3A%2Ftags%2Fwiki-test-spec%22%2C%22text%22%3A%22title%3A%20Output%5Cn%5Cn%5C%5Cwhitespace%20trim%5Cn%5C%5Cprocedure%20all.identity%28%29%20%20%20%20%20%20%20%20%20%20%3Aall%5Ball%5B%5D%5D%5Cn%5C%5Cprocedure%20and.identity%28%29%20%20%20%20%20%20%20%20%20%20%3Aand%5Ball%5B%5D%5D%5Cn%5C%5Cprocedure%20cascade.identity%28%29%20%20%20%20%20%20%3Acascade%5Ball%5B%5D%5D%5Cn%5C%5Cprocedure%20else.one%28%29%20%20%20%20%20%20%20%20%20%20%20%20%20%20%3Aelse%5B%5B1%5D%5D%5Cn%5C%5Cprocedure%20except.two%28%29%20%20%20%20%20%20%20%20%20%20%20%20%3Aexcept%5B%5B2%5D%5D%5Cn%5C%5Cprocedure%20filter.identity%28%29%20%20%20%20%20%20%20%3Afilter%5Ball%5B%5D%5D%5Cn%5C%5Cprocedure%20intersection.identity%28%29%20%3Aintersection%5Ball%5B%5D%5D%5Cn%5C%5Cprocedure%20map.identity%28%29%20%20%20%20%20%20%20%20%20%20%3Amap%5Ball%5B%5D%5D%5Cn%5C%5Cprocedure%20or.identity%28%29%20%20%20%20%20%20%20%20%20%20%20%3Aor%5Ball%5B%5D%5D%5Cn%5C%5Cprocedure%20reduce.identity%28%29%20%20%20%20%20%20%20%3Areduce%5Ball%5B%5D%5D%5Cn%5C%5Cprocedure%20sort.identity%28%29%20%20%20%20%20%20%20%20%20%3Asort%5Ball%5B%5D%5D%5Cn%5C%5Cprocedure%20then.identity%28%29%20%20%20%20%20%20%20%20%20%3Athen%5Ball%5B%5D%5D%5Cn%5Cn%3C%24text%20text%3D%7B%7B%7B%20%5Cn%20%20%5B%5Ball.identity%5Dsubfilter%3Call.identity%3E%5D%5Cn%20%20%5B%5Band.identity%5Dsubfilter%3Cand.identity%3E%5D%5Cn%20%20%5B%5Bcascade.identity%5Dsubfilter%3Ccascade.identity%3E%5D%5Cn%20%20%5B%5Belse.one%5Dsubfilter%3Celse.one%3E%5D%5Cn%20%20%5Benlist%5B2%20except.two%5Dsubfilter%3Cexcept.two%3E%5D%5Cn%20%20%5B%5Bfilter.identity%5Dsubfilter%3Cfilter.identity%3E%5D%5Cn%20%20%5B%5Bintersection.identity%5Dsubfilter%3Cintersection.identity%3E%5D%5Cn%20%20%5B%5Bmap.identity%5Dsubfilter%3Cmap.identity%3E%5D%5Cn%20%20%5B%5Bor.identity%5Dsubfilter%3Cor.identity%3E%5D%5Cn%20%20%5B%5Breduce.identity%5Dsubfilter%3Creduce.identity%3E%5D%5Cn%20%20%5B%5Bsort.identity%5Dsubfilter%3Csort.identity%3E%5D%5Cn%20%20%5B%5Bthen.identity%5Dsubfilter%3Cthen.identity%3E%5D%5Cn%20%20%3Aand%5Bjoin%5B%20%5D%5D%5Cn%7D%7D%7D%2F%3E%5Cn%5Cn%2B%5Cntitle%3A%20ExpectedResult%5Cn%5Cnall.identity%201%20or.identity%22%7D%5D)

Edit: Added explicit call to the `title` operator to hopefully make it a little more clear.

**Comment by pmario** [comment]
I am not sure what you want to show with your testcases. 

There is the [function-operator](https://tiddlywiki.com/#function%20Operator), which allows us to call functions in filter-runs

>[New in v5.3.0](https://tiddlywiki.com/#Release%205.3.0) The function operator applies a named [function](https://tiddlywiki.com/#Functions) to the input titles, and returns the results from the function. The function is invoked once with all of the input titles (in contrast, the [filter Operator](https://tiddlywiki.com/#filter%20Operator) invokes its function separately for each input title).

If you want to call a function with inputs, you have to give it input parameters

```
\function test-match(var1, var2) [<var1>match[one]else<var2>] 

{{{ [function[test-match],[one],[two]]  }}}

{{{ [function[test-match],[oneXX],[two]]  }}}
```

Calling functions without the `function[]` operator is possible, if you include a **dot in the function name**. We could call that a shortcut.

```
\function test.match(var1, var2) [<var1>match[one]else<var2>] 

{{{ [<test.match one two>]  }}}

{{{ [<test.match oneXX two>]  }}}
```

But that's probably not what you want to show us.

**Reply by btheado** [reply]
Thanks, @pmario. In trying to demonstrate every filter prefix, I see I made it confusing.

I'll try again.

```
\function map.identity1()          [all[]] :map[all[]]
\function map.identity2()                  :map[all[]]

<$text text={{{ [title[abc]map.identity1[]] }}}/> - <$text text={{{ [title[abc]map.identity2[]] }}}/>
```

My question is why is the output of the above `abc -` instead of `abc - abc`. Why isn't the input supplied to the `map.identity2` function?

I see the same behavior for all the other filter run prefixes except for `:all` and `:or`.

**Comment by Jermolene** [comment]
Thanks @btheado, very intriguing. I would also think that `[all[]]` should be a no-op in this case. At the moment I'm baffled; the implementation of "[all[]]" is very simple, and it's really hard to see how this can be happening. I'll try coming back to it with fresh eyes.

I did create a simple test case:

```
title: Filters/FunctionInput
description: Verify expected inputs to functions
type: text/vnd.tiddlywiki-multiple
tags: [[$:/tags/wiki-test-spec]]

title: Output

\whitespace trim
\function map.identity1()          [all[]] :map[all[]]
\function map.identity2()                  :map[all[]]

<$text text={{{ [title[abc]map.identity1[]] }}}/> - <$text text={{{ [title[abc]map.identity2[]] }}}/>

<$text text={{{ [title[abc]function[map.identity1],[]] }}}/> - <$text text={{{ [title[abc]function[map.identity2],[]] }}}/>
+
title: ExpectedResult

<p>abc-abc</p><p>abc-abc</p>
```

**Reply by btheado** [reply]
>  the implementation of "[all[]]" is very simple

Right, it doesn't seem to have anything to do with the operator. You could use any "no-op" operator in its place such as `[prefix[]]`

The filter run prefix is what matters. Since `[all[]]` is an implicit `:or` and `:or` is one of the two filter prefixes for which it works. You can replace map.identity1 with

```
\function map.identity1()          :all[all[]] :map[all[]]
```

and it still gives the `abc` output.

**Comment by btheado** [comment]
Here is my analysis of the code. I don' t have a good big picture understanding of `source` and `results`, but the reason functions which begin with filter run prefixes such as `:map` don't see the input is because the `results` are not passed along to the function's "subfilter". When `results` start empty for these filter runs, the `source` is ignored. Filter run prefixes such as `:all` and `:or` don't care that the `results` start empty. The `source` will be processed just fine.

The `function` operator calls `getVariableInfo`, passing `source` in the options parameter: 
https://github.com/TiddlyWiki/TiddlyWiki5/blob/028c80782d105beb90f5d58a7f22e865c7e8c6f4/core/modules/filters/function.js#L24-L26

`getVariableInfo` calls `this.wiki.filterTiddlers`. The `options.source` is passed along:
https://github.com/TiddlyWiki/TiddlyWiki5/blob/028c80782d105beb90f5d58a7f22e865c7e8c6f4/core/modules/widgets/widget.js#L164

Each filter run always starts with empty `results` due to this code:
https://github.com/TiddlyWiki/TiddlyWiki5/blob/028c80782d105beb90f5d58a7f22e865c7e8c6f4/core/modules/filters.js#L344-L349

The `:all` filter run prefix (`:or` is similar) doesn't care the input `results` are empty...it will directly use `source` to get the input values. Therefore, these filter runs are able to see the input:
https://github.com/TiddlyWiki/TiddlyWiki5/blob/028c80782d105beb90f5d58a7f22e865c7e8c6f4/core/modules/filterrunprefixes/all.js#L19-L22

However, the `:map` filter run prefix and most of the other filter run prefixes explicitly return early when empty input `results` are encountered.  Therefore these filter runs are NOT able to see the input:
https://github.com/TiddlyWiki/TiddlyWiki5/blob/028c80782d105beb90f5d58a7f22e865c7e8c6f4/core/modules/filterrunprefixes/map.js#L15-L19

The `:and` filter doesn't have a check for empty `results`, but it does ignore the input `source` and generates a new one from `results.toArray()`. Since since `results` is empty, this newly generated `source` doesn't give any output. This filter run is also unable to see the input:
https://github.com/TiddlyWiki/TiddlyWiki5/blob/028c80782d105beb90f5d58a7f22e865c7e8c6f4/core/modules/filterrunprefixes/and.js#L19-L24

Edit: Link to the correct filter run results initialization

**Reply by Jermolene** [reply]
Thank you @btheado. So it seems like we have definite bugs in the `map` and `and` filter run prefixes; this would appear not to just be a documentation issue. It would be interesting to see if there was any unexpected impact if they were to be fixed.

**Reply by btheado** [reply]
I made this horrible, hacky change to automatically insert `[all[]]` at the start of any filter for which source is not empty and the first filter run is not a `:all` or `:or` filter run

```diff
diff --git a/core/modules/filters.js b/core/modules/filters.js
index aa82a352a..3cf3a0563 100644
--- a/core/modules/filters.js
+++ b/core/modules/filters.js
@@ -216,6 +216,15 @@ exports.getFilterRunPrefixes = function() {
 
 exports.filterTiddlers = function(filterString,widget,source) {
        var fn = this.compileFilter(filterString);
+       if (source) {
+               var filterParseTree = this.parseFilter(filterString);
+               if (filterParseTree.length > 0) {
+                       var firstRun = filterParseTree[0];
+                       if (firstRun.prefix != "" && firstRun.namedPrefix != "or" && firstRun.prefix != "=" && firstRun.namedPrefix != "all") {
+                               fn = this.compileFilter("[all[]]" + filterString);
+                       }
+               }
+       }
        return fn.call(this,source,widget);
 };
```

> It would be interesting to see if there was any unexpected impact if they were to be fixed.

All tests pass, but only the tests I provided in OP actually exercise this new code.

It is hard for me to imagine someone having code which depends on something like `\function map.identity() :map[all[]]` returning empty results, but maybe I need to stretch my imagination :-).

I wonder if @saqimtiaz has any input here. I have a vague recollection that he recommended here or on the forums about inserting the `[all[]]` at the start of a function in order to get input fed to `:map`. I doubt I would have come up with this `[all[]]` workaround without that vague recollection. I can't seem to find reference to his recommendation, though.

I'm not sure what the "proper" fix corresponding to the above would look like. I need to learn more about the code in filters.js.

**Comment by Jermolene** [comment]
Linking to a discussion about the same issue: https://talk.tiddlywiki.org/t/wikitext-code-editor/11211

**Reply by btheado** [reply]
That discussion appears unrelated. Maybe you meant this one: https://talk.tiddlywiki.org/t/is-this-a-bug-cant-directly-use-prefixed-filter-run-in-function/11313

**Reply by Jermolene** [reply]
Thanks @btheado my mistake.

**Comment by btheado** [comment]
I found this code which seems to be a workaround that is somewhat related:
https://github.com/TiddlyWiki/TiddlyWiki5/blob/1f34c628288cded3c322d56a345de924e6b1b257/core/modules/widgets/action-listops.js#L72-L76

It was introduced in #6540. @Jermolene, it looks like you added `this.wiki.makeTiddlerIterator(inputList)` as the source parameter to the `this.wiki.filterTiddlers` call. In the same commit you automatically prepend `"[all[]] "` to the subfilter.

Maybe you already knew `[all[]]` prefix would be required since most filter run prefixes ignore the `source` when the results linked list is empty. Or maybe you discovered that by trial and error. It is hard to tell from the history.

**Reply by Jermolene** [reply]
Hi @btheado I think this might be an accidental similarity, depending on how one looks at it. The approach taken by #6540 is to try to retain backwards compatibility with the existing code while avoiding using dangerous stringified list construction. The code looks like that because there is no API for passing a list of titles as the starting accumulator of a filter, which is of course the peculiarity we have been discussing.
